src/core/write_list.cc

```diff
@@ -44,102 +44,106 @@ THE SOFTWARE.
 
 namespace core {
 
-void Cons_O::__write__(T_sp stream) const {
-  T_sp x;
-  bool circle;
-  Fixnum print_level, print_length;
-  _Index i;
-  T_sp y;
-  SYMBOL_EXPORT_SC_(CorePkg, _SHARP__BANG_);
-  if (this->ocar() == _sym__SHARP__BANG_) {
-    clasp_write_string("#!", stream);
-    x = this->cdr();
-    write_object(x, stream);
-    return;
+bool Cons_O::maybe_write_quoted_form(bool tail, T_sp stream) const {
+  if (car().nilp() || !cdr().consp() || oCdr(cdr()).notnilp())
+    return false;
+
+  T_sp op = car();
+
+  if (op == cl::_sym_quote) {
+    if (tail)
+      return false;
+    clasp_write_char('\'', stream);
+    write_object(oCar(cdr()), stream);
+    return true;
   }
-  if ((this->cdr()).consp() && oCdr(this->cdr()).nilp()) {
-    if (this->ocar() == cl::_sym_quote) {
-      clasp_write_char('\'', stream);
-      x = oCar(this->cdr());
-      write_object(x, stream);
-      return;
-    }
-    if (this->ocar() == cl::_sym_function) {
-      clasp_write_char('#', stream);
-      clasp_write_char('\'', stream);
-      x = oCar(this->cdr());
-      write_object(x, stream);
-      return;
-    }
-    if (this->ocar() == _sym_quasiquote) {
-      clasp_write_char('`', stream);
-      x = oCar(this->cdr());
-      write_object(x, stream);
-      return;
-    }
-    if (this->ocar() == _sym_unquote) {
-      clasp_write_char(',', stream);
-      x = oCar(this->cdr());
-      write_object(x, stream);
-      return;
-    }
-    if (this->ocar() == _sym_unquote_splice) {
-      clasp_write_string(",@", stream);
-      x = oCar(this->cdr());
-      write_object(x, stream);
-      return;
-    }
-    if (this->ocar() == _sym_unquote_nsplice) {
-      clasp_write_string(",.", stream);
-      x = oCar(this->cdr());
-      write_object(x, stream);
-      return;
-    }
+
+  if (op == cl::_sym_function) {
+    if (tail)
+      return false;
+    clasp_write_string("#'", stream);
+    write_object(oCar(cdr()), stream);
+    return true;
+  }
+
+  T_sp quasiquote = _sym_STARquasiquoteSTAR->symbolValue();
```

Comment:
Might be worth explaining what `*quasiquote*` is for, somewhere, briefly. If I've divined it correctly it's not that complicated - a plist from streams to a "are we currently writing a quasiquote form" marker - but it's nontrivial. Also, does it do okay with nested quasiquotation? As written it looks like printing `(core:unquote foo)` outside of quasiquotation will just print that instead of `,foo`, but how about ``(`(foo `(bar ,,,baz)))`` - will that print with an extra comma even though it has one too many unquotes, instead of as ``(`(foo `(bar (core:unquote ,,baz)))``? (Is this too pedantic to care about?)

Reply:
Correct, it will handling nesting like that. I added a few regression tests that show a few levels of that.

Reply:
Specifically, because getf returns the first key/value binding as `(list* stream nil core:*quasiquote*)` will ignore the trailing nestings.